// appygram/lib/appygram.js
// Generated by CoffeeScript 1.3.3
(function() {
  var Appygram, Singleton, request,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Singleton = require('singleton');

  request = require('request');

  Appygram = (function(_super) {

    __extends(Appygram, _super);

    function Appygram() {
      this.defaults();
    }

    Appygram.prototype.defaults = function() {
      this.api_key = void 0;
      this.endpoint = 'https://arecibo.appygram.com/traces';
      this.version = JSON.parse((require('fs')).readFileSync(__dirname + '/../package.json')).version;
      this.user_location = 'user';
      this.include_user = false;
      return this.app_name = "node application";
    };

    Appygram.prototype.setApiKey = function(api_key) {
      return this.api_key = api_key;
    };

    Appygram.prototype.reset_to_default = function() {
      return this.defaults();
    };

    Appygram.prototype.errorHandler = function(err, res, req, next) {
      var appy, error, options, params;
      appy = Appygram.get();
      if (appy.api_key === void 0) {
        return console.error('Please define Appygram\'s API_KEY with appygram.api_key = \'your_api_key\' or appygram.setApiKey(\'your_api_key\')\
        \ If you need an API key please visit http://www.appygram.com/dashboard and request one for your project.');
      } else {
        error = new Error(err);
        params = {
          api_key: appy.api_key,
          name: "Exception",
          topic: "Exception",
          message: error.toString(),
          trace: JSON.stringify({
            "class": error.message,
            message: error.message,
            backtrace: appy.parseBackTrace(error.stack)
          }),
          platform: "appygram-node" + appy.version,
          software: appy.app_name,
          app_json: appy.include_user && (req[appy.user_location] != null) ? {
            user: JSON.stringify(req[appy.user_location])
          } : void 0
        };
        options = {
          uri: appy.endpoint,
          method: 'POST',
          body: JSON.stringify(params),
          json: true,
          headers: {
            "User-Agent": "appygram-node/" + appy.version,
            "Accept": "application/json",
            "Content-Type": "application/json"
          }
        };
        request(options, function(error, response, body) {
          console.log('Processed appygram');
          if (appy.debug) {
            return next(err);
          }
        });
        if ((next != null) && !appy.debug) {
          return next(err);
        }
      }
    };

    Appygram.prototype.parseBackTrace = function(trace) {
      var list, msgs, rest, traceList, traces, _i, _len;
      list = trace.split('\n');
      traces = [];
      for (_i = 0, _len = list.length; _i < _len; _i++) {
        trace = list[_i];
        traceList = [];
        msgs = trace.split('(');
        if (msgs.length > 1) {
          traceList.push(msgs[0]);
          rest = msgs[1].split(':');
          traceList.push(rest[0]);
          traceList.push(rest[1]);
        }
        if (traceList.length > 0) {
          traces.push(traceList);
        }
      }
      return traces;
    };

    return Appygram;

  })(Singleton);

  module.exports = Appygram.get();

}).call(this);
